### ISSUE TITLE
120hz 144hz bad performance

### ISSUE THREAD

**Issue by phisch** [gnif/LookingGlass#98]
### Required information

Host CPU: i7-5820K
Host GPU: GTX970
Guest GPU: GTX970
Host Kernel version: 4.18.5
Host QEMU version: 3.0.0

When using 120hz or 144hz, LookingGlass is rendering with stutters. 

With the default configuration and opengl vsync enabled, the FPS are at ~120 when the UPS are at around ~1, once the UPS rise (e.g. through moving the mouse or a window) the FPS go down to ~60 (like i would expect from vsync).

So the first question here would be, why are the FPS sitting at ~120 when there is nothing to draw?

Once opengl vsync is disabled, the FPS sit at the FPS limit (200 by default) and drop down between 110 and 130 once the UPS rise to about ~120. Even if the FPS are hitting 120, the window is rendered with quite some stutters, it looks like it's drawn with 60 FPS. I can hardly - if at all - see a difference between 120hz and 60hz.

Just to clarify, both host and guest are correctly set to 120 or 140hz (depending on what i was testing).

I tested this with both the latest release and the bleeding edge, both have the same result for me.

Let me know if i can provide any more information, logs or test-results! (love the project btw.)

TLDR: Drawn FPS are too high when there is nothing to draw, and 120hz and 144hz look like 60hz.

**Comment by mattfredwill** [comment]
According to gnif on the [forum thread](https://forum.level1techs.com/t/looking-glass-triage/130952/1139):

> >Also I will be using a 144hz 2560x1080 display.

>This is beyond LG’s capability at this time, it will work but your FPS will suffer.

This lines up with my experience as well (with higher res rather than higher hz, but the problem is the same)

**Comment by phisch** [comment]
I have tried lowering the resolution to 1080p, which gave me ~144 UPS and plenty (~190 FPS), still the result is the same. The rendering is way smoother on the directly connected monitor than in the rendered window from looking glass.

Also i noticed that the image quality is slightly better on the directly connected monitor, but that might be due the differences in how windows and linux (or probably opengl) handle scaling differently. (this was only for 1080p resolution on 1440p monitors)

I really doubt this is a performance issue, there is something different going on.

**Comment by phisch** [comment]
Checked out the client, and i can't find anything that would cause an issue here, it renders whatever the "host" sends as expected.

I would assume that there is something fishy going on with the host and how it reads the framebuffer with DXGI. Might try to debug this later when i have some more time.

**Comment by Lessaj** [comment]
Are you on Wayland or Xorg?

**Comment by phisch** [comment]
Xorg

**Comment by Programming4life** [comment]
I have this issue too. Running on a 1080p 144hz monitor, Looking Glass has slight stutters. Changing my display to 60hz is alot smoother vs ruining Looking Glass on a 144hz monitor.

**Comment by gnif** [comment]
This is not the place for this, GitHub issues are for bug reports, not discussons on performance. Keep these discussions to the L1Tech forums please.

**Comment by gnif** [comment]
I am sorry but you think wrong, I can tell you with 100% certainty that the micro stuttering you are seeing is due to desync, even if you were running at one frame per second, clock drift would eventually see a frame drop due to the frame arriving too late. The higher the resolution and refresh, the more frequenty this occurs.

**Comment by svenstaro** [comment]
Shouldn't the general issue here (bad performance at higher refresh rates) still be considered an issue that should eventually be fixed, though?

**Comment by gnif** [comment]
If you want to design your own GPU that supports sync lock, sure... there is no solution to this short of a custom GPU.

**Comment by svenstaro** [comment]
Alright, so it's a hard hardware limitation as opposed to an implementation thing, and using something like the NVIDIA Streaming API instead of DXGI wouldn't work either then. Thanks for clearing that up.

**Comment by gnif** [comment]
Just so everyone here knows, this issue is largely resolved through the hundreds of changes over the last 2 years, most notably by the addition of a guest disciplined FPS in the client and the discovery of how to get windows to give the capture thread priority over the running application's renderer. There are still instances where micro stutters can happen, but this comes down to clock drift as mentioned prior, however it is far less noticeable now.
